From fbb444688c95fbb7c516dfe8fdd4622bbae734e5 Mon Sep 17 00:00:00 2001
From: Zhihao Cheng <chengzhihao1@huawei.com>
Date: Thu, 13 Jul 2023 16:00:26 +0800
Subject: [PATCH 20/36] [Huawei] Revert "[Huawei] ext4: Stop trying writing
 pages if no free blocks generated"

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188972

--------------------------------

This reverts commit 366247c69118429c524d1d0535d30142cd0317b8.

When ext4 runs out of space, there could be a potential data lost in
ext4_writepages:
If there are many preallocated blocks for some files, e4b bitmap is
different from block bitmap, and there are more free blocks accounted
by block bitmap.

     ext4_writepages                         P2
ext4_mb_new_blocks                      ext4_map_blocks
 ext4_mb_regular_allocator // No free bits in e4b bitmap
 ext4_mb_discard_preallocations_should_retry
  ext4_mb_discard_preallocations
   ext4_mb_discard_group_preallocations
    ext4_mb_release_inode_pa // updates e4b bitmap by pa->pa_free
     mb_free_blocks
                                         ext4_mb_new_blocks
                                          ext4_mb_regular_allocator
					 // Got e4b bitmap's free bits
 ext4_mb_regular_allocator  // After 3 times retrying, ret ENOSPC

ext4_writepages
 mpage_map_and_submit_extent
  mpage_map_one_extent // ret ENOSPC
  if (err == -ENOSPC && EXT4_SB(sb)->s_mb_free_pending)
  // s_mb_free_pending is 0
  *give_up_on_write = true  // Abandon writeback, data lost!

Fixes: 366247c691184 ("[Huawei] ext4: Stop trying writing pages if ...")
Signed-off-by: Zhihao Cheng <chengzhihao1@huawei.com>
---
 fs/ext4/inode.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fs/ext4/inode.c b/fs/ext4/inode.c
index 388b13f37bd3..ed2c466ad891 100644
--- a/fs/ext4/inode.c
+++ b/fs/ext4/inode.c
@@ -2478,7 +2478,7 @@ static int mpage_map_and_submit_extent(handle_t *handle,
 			 * is non-zero, a commit should free up blocks.
 			 */
 			if ((err == -ENOMEM) ||
-			    (err == -ENOSPC && EXT4_SB(sb)->s_mb_free_pending)) {
+			    (err == -ENOSPC && ext4_count_free_clusters(sb))) {
 				if (progress)
 					goto update_disksize;
 				return err;
-- 
2.17.1

